numpy/core/shape_base.py

```diff
@@ -429,42 +435,157 @@ def _block_check_depths_match(arrays, parent_index=[]):
             # propagate our flag that indicates an empty list at the bottom
             if index[-1] is None:
                 first_index = index
-        return first_index, max_arr_ndim
+
+        return first_index, max_arr_ndim, n_copied
     elif type(arrays) is list and len(arrays) == 0:
         # We've 'bottomed out' on an empty list
-        return parent_index + [None], 0
+        return parent_index + [None], 0, 0
     else:
         # We've 'bottomed out' - arrays is either a scalar or an array
-        return parent_index, _nx.ndim(arrays)
+        n_copied = _nx.size(arrays)
+        return parent_index, _nx.ndim(arrays), n_copied
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
+    return accumulated
+
+
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape and slices prefixes.
+
+    These help in nested concatation.
+    Returns
+    -------
+    shape: tuple of int
+        This tuple satisfies:
+        ``shape, slice_prefixes = _concatenate_shapes([arr.shape for shape in arrs], axis)``
+        ``shape == concatenate(arrs, axis).shape``
+
+    slice_prefixes: tuple of (slice(start, end), )
+        Produce the prefixes that are required for nested slicing.
+
+        Returned slices are tuples optimized for being concatenated with other
+        tuples of slices.
 
+        # TODO: I need to reword this comment
+        For the following situation::
+            ret = concatenate([a, b, c], axis=0)
+            sl_a, sl_b, sl_c = concatenate_slices([a, b, c], axis=0)
+        The result will be that ``ret[sl_a] == a``, ``ret[sl_b] == b``, and
+        ``ret[sl_c] == c``.
+        Similar properties hold for other values of ``axis``.
 
-def _block(arrays, max_depth, result_ndim):
     """
-    Internal implementation of block. `arrays` is the argument passed to
+    # Cache a result that will be reused.
+    shape_at_axis = [shape[axis] for shape in shapes]
+
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    first_shape_pre = first_shape[:axis]
+
+    # Why do I not need to check shape[axis+1:] == first_shaoe[axis+1:]?
+    # Avoid this check makes things faster
+    # empty lists are falsy
+    if [True for shape in shapes if shape[:axis] != first_shape_pre]:
```

Comment:
What about `any(shape[:axis] != first_shape_pre for shape in shapes):` ?

Reply:
I'll go back to `any`, it is definitely cleaner.